hls-plugin-api/src/Ide/Plugin/Properties.hs

```diff
@@ -0,0 +1,341 @@
+{-# LANGUAGE AllowAmbiguousTypes  #-}
+{-# LANGUAGE ConstraintKinds      #-}
+{-# LANGUAGE DataKinds            #-}
+{-# LANGUAGE FlexibleContexts     #-}
+{-# LANGUAGE FlexibleInstances    #-}
+{-# LANGUAGE GADTs                #-}
+{-# LANGUAGE KindSignatures       #-}
+{-# LANGUAGE OverloadedStrings    #-}
+{-# LANGUAGE RecordWildCards      #-}
+{-# LANGUAGE ScopedTypeVariables  #-}
+{-# LANGUAGE TypeApplications     #-}
+{-# LANGUAGE TypeFamilies         #-}
+{-# LANGUAGE TypeOperators        #-}
+{-# LANGUAGE UndecidableInstances #-}
+-- See Note [Constraints]
+{-# OPTIONS_GHC -Wno-redundant-constraints #-}
+
+module Ide.Plugin.Properties
+  ( PropertyType (..),
+    ToHsType,
+    MetaData (..),
+    PropertyKey,
+    SPropertyKey (..),
+    Properties,
+    HasProperty,
+    emptyProperties,
+    defineNumberProperty,
+    defineStringProperty,
+    defineBooleanProperty,
+    defineObjectProperty,
+    defineArrayProperty,
+    defineEnumProperty,
+    toDefaultJSON,
+    toVSCodeExtensionSchema,
+    usePropertyEither,
+    useProperty,
+    (&),
+  )
+where
+
+import qualified Data.Aeson       as A
+import qualified Data.Aeson.Types as A
+import           Data.Either      (fromRight)
+import           Data.Function    ((&))
+import           Data.Kind        (Constraint)
+import qualified Data.Map.Strict  as Map
+import           Data.Proxy       (Proxy (..))
+import qualified Data.Text        as T
+import           GHC.TypeLits
+import           Unsafe.Coerce    (unsafeCoerce)
+
+-- ---------------------------------------------------------------------
+
+data PropertyType
+  = TNumber
+  | TString
+  | TBoolean
+  | TObject
+  | TArray
+  | TEnum
+
+type family ToHsType (t :: PropertyType) where
+  ToHsType 'TNumber = Int
+  ToHsType 'TString = T.Text
+  ToHsType 'TBoolean = Bool
+  ToHsType 'TObject = A.Object
+  ToHsType 'TArray = A.Array
+  ToHsType 'TEnum = T.Text -- supports only text enum now
+
+-- ---------------------------------------------------------------------
+
+data MetaData (t :: PropertyType) where
+  MetaData ::
+    (IsNotTEnum t) =>
```

Comment:
Is this constraint just for type errors, or do you use it as evidence somehow?

Reply:
Now it's just for type errors. The idea is to let enum properties have unique metadata that includes `enumValues`. Perhaps there's a better way?

Reply:
Ah,

```haskell
type family IsTEnum (t :: PropertyType) :: Bool where
  IsTEnum 'TEnum = 'True
  IsTEnum _ = 'False
```

is better. We can use `(IsTEnum t ~ 'True)` and `(IsTEnum t ~ 'False)` as the evidence at call sites, which also avoids that ghc warns that an impossible case is missing in pattern matching.